In [3]:
# Filtering out the warnings

import warnings

warnings.filterwarnings('ignore')

In [4]:
# Importing the required libraries

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

1) Import Data from csv files 

In [5]:
credits = pd.read_csv('credits.csv')

In [6]:
credits.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

In [7]:
keywords = pd.read_csv('keywords.csv')

In [8]:
keywords.head()

id                                           keywords
0    862  [{'id': 931, 'name': 'jealousy'}, {'id': 4290,...
1   8844  [{'id': 10090, 'name': 'board game'}, {'id': 1...
2  15602  [{'id': 1495, 'name': 'fishing'}, {'id': 12392...
3  31357  [{'id': 818, 'name': 'based on novel'}, {'id':...
4  11862  [{'id': 1009, 'name': 'baby'}, {'id': 1599, 'n...

In [9]:
links = pd.read_csv('links.csv')

In [10]:
links.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [202]:
md = pd.read_csv('movies_metadata.csv')

In [203]:
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [13]:
rating = pd.read_csv('ratings.csv')

In [14]:
rating.head()

userId  movieId  rating   timestamp
0       1      110     1.0  1425941529
1       1      147     4.5  1425942435
2       1      858     5.0  1425941523
3       1     1221     5.0  1425941546
4       1     1246     5.0  1425941556

Analyzing the Dataset 

In [15]:
from tabulate import tabulate

In [16]:
table = [['DATASET','Rows and Column'], 
         ['credits',credits.shape],['keywords',keywords.shape],['links',links.shape],['movies medata',md.shape],['rating',rating.shape]]

In [17]:
print(tabulate(table, headers='firstrow',tablefmt='fancy_grid'))

╒═══════════════╤═══════════════════╕
│ DATASET       │ Rows and Column   │
╞═══════════════╪═══════════════════╡
│ credits       │ (45476, 3)        │
├───────────────┼───────────────────┤
│ keywords      │ (46419, 2)        │
├───────────────┼───────────────────┤
│ links         │ (45843, 3)        │
├───────────────┼───────────────────┤
│ movies medata │ (45466, 24)       │
├───────────────┼───────────────────┤
│ rating        │ (26024289, 4)     │
╘═══════════════╧═══════════════════╛


In [18]:
print ("Credits Dataset with Data Type")
credits.info()

Credits Dataset with Data Type
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45476 entries, 0 to 45475
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   cast    45476 non-null  object
 1   crew    45476 non-null  object
 2   id      45476 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 1.0+ MB


In [19]:
print("links Dataset with Data Type")
links.info()

links Dataset with Data Type
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45843 entries, 0 to 45842
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   movieId  45843 non-null  int64  
 1   imdbId   45843 non-null  int64  
 2   tmdbId   45624 non-null  float64
dtypes: float64(1), int64(2)
memory usage: 1.0 MB


In [20]:
print("movies metadata Dataset with Data Type")
md.info()

movies metadata Dataset with Data Type
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45466 entries, 0 to 45465
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  45466 non-null  object 
 1   belongs_to_collection  4494 non-null   object 
 2   budget                 45466 non-null  object 
 3   genres                 45466 non-null  object 
 4   homepage               7782 non-null   object 
 5   id                     45466 non-null  object 
 6   imdb_id                45449 non-null  object 
 7   original_language      45455 non-null  object 
 8   original_title         45466 non-null  object 
 9   overview               44512 non-null  object 
 10  popularity             45461 non-null  object 
 11  poster_path            45080 non-null  object 
 12  production_companies   45463 non-null  object 
 13  production_countries   45463 non-null  object 
 14  release_date   

In [21]:
print("Rating Dataset with Data Type")
rating.info()

Rating Dataset with Data Type
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26024289 entries, 0 to 26024288
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 794.2 MB


In [22]:
print("Keyword Dataset with Data Type")
rating.info()

Keyword Dataset with Data Type
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26024289 entries, 0 to 26024288
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 794.2 MB


DATA CLEANUP 

     a) MOVIE METADATA (md file)

In [23]:
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview  ... release_date  \
0  Led by Woody, Andy's toys live happily in his ...  ...   1995-10-30   
1  When siblings Judy and Peter discover an encha...  ...   1995-12-15   
2  A family wedding reignites the ancient feud be...  ...   1995-12-22   
3  Cheated on, mistreated and stepped on, the wom...  ...   1995-12-22   
4  Just when George Banks has recovered from his ...  ...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                                            tagline  \
0  Released                                                NaN   
1  Released          Roll the dice and unleash the excitement!   
2  Released  Still Yelling. Still Fighting. Still Ready for...   
3  Released  Friends are the people who let you be yourself...   
4  Released  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  
0                    Toy Story  False          7.7     5415.0  
1                      Jumanji  False          6.9     2413.0  
2             Grumpier Old Men  False          6.5       92.0  
3            Waiting to Exhale  False          6.1       34.0  
4  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 24 columns]

In [24]:
md.loc[0]

adult                                                                False
belongs_to_collection    {'id': 10194, 'name': 'Toy Story Collection', ...
budget                                                            30000000
genres                   [{'id': 16, 'name': 'Animation'}, {'id': 35, '...
homepage                              http://toystory.disney.com/toy-story
id                                                                     862
imdb_id                                                          tt0114709
original_language                                                       en
original_title                                                   Toy Story
overview                 Led by Woody, Andy's toys live happily in his ...
popularity                                                       21.946943
poster_path                               /rhIRbceoE9lR4veEXuwCC2wARtG.jpg
production_companies        [{'name': 'Pixar Animation Studios', 'id': 3}]
production_countries     

We need to work on cleaning the data of belongs_to_collection and genres

In [25]:
# Drop unused column 

md = md.drop(['adult'], axis=1)
md = md.drop(['tagline'],axis=1)

### Preview Data 
md.head()

belongs_to_collection    budget  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1                                                NaN  65000000   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3                                                NaN  16000000   
4  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview popularity  ...  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
1  When siblings Judy and Peter discover an encha...  17.015539  ...   
2  A family wedding reignites the ancient feud be...    11.7129  ...   
3  Cheated on, mistreated and stepped on, the wom...   3.859495  ...   
4  Just when George Banks has recovered from his ...   8.387519  ...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                        title  video vote_average vote_count  
0  Released                    Toy Story  False          7.7     5415.0  
1  Released                      Jumanji  False          6.9     2413.0  
2  Released             Grumpier Old Men  False          6.5       92.0  
3  Released            Waiting to Exhale  False          6.1       34.0  
4  Released  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 22 columns]

PRE-PROCESSING

a) Check Missing Values 

In [26]:
print("Missing Value of Movie Dataset\n")
print( md.isna().sum())
print("\nMissing Value of Credits Dataset\n")
print (links.isna().sum())
print("\nMissing Value of Keywords Dataset\n")
print( keywords.isna().sum())
print("\nMissing Value of links Dataset\n")
print( links.isna().sum())
print("\nMissing Value of Rating Dataset\n")
print( rating.isna().sum())

Missing Value of Movie Dataset

belongs_to_collection    40972
budget                       0
genres                       0
homepage                 37684
id                           0
imdb_id                     17
original_language           11
original_title               0
overview                   954
popularity                   5
poster_path                386
production_companies         3
production_countries         3
release_date                87
revenue                      6
runtime                    263
spoken_languages             6
status                      87
title                        6
video                        6
vote_average                 6
vote_count                   6
dtype: int64

Missing Value of Credits Dataset

movieId      0
imdbId       0
tmdbId     219
dtype: int64

Missing Value of Keywords Dataset

id          0
keywords    0
dtype: int64

Missing Value of links Dataset

movieId      0
imdbId       0
tmdbId     219
dtype: int64

Missing Val

MODEL 1 : Simple Recommodation 

As a first step, let's calculate the value of C, the mean rating across all movies

In [29]:
## Calculate the mean of Vote Average 

C = md['vote_average'].mean()
print(C)


5.618207215133889

 The average rating of a movie on IMDB is around 5.6 on a scale of 10.

We will use 90th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 90% of the movies in the list.

In [30]:
# Calculate the minimum number of votes required to be in the chart, m
m = md['vote_count'].quantile(0.90)
print(m)

160.0


In [31]:
## Filter out qualified movie for the set 

q_movies = md.copy().loc[md['vote_count'] >= m ]
q_movies.shape

(4555, 22)

It is clear that there are around 10% movies with vote count more than 160.Now, we need to calculate our metric for each qualified movie. To do this, we will define a function, weighted_rating() and define a new feature score, of which we'll calculate the value by applying this function to our DataFrame of qualified movies:

In [33]:
def weighted_rating(x, m=m, C=C) : 
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

In [34]:
# Define a new feature 'score' and calculate its value with `weighted_rating()`
q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

Finally, let's sort the DataFrame based on the score feature and output the title, vote count, vote average and weighted rating or score of the top 20 movies.

In [205]:
#Sort movies based on score calculated above
q_movies = q_movies.sort_values('score', ascending=False)

#Print the top 15 movies
q_movies[['title', 'vote_count', 'vote_average', 'score']].head(15)

title  vote_count  vote_average     score
314           The Shawshank Redemption      8358.0           8.5  8.445869
834                      The Godfather      6024.0           8.5  8.425439
10309      Dilwale Dulhania Le Jayenge       661.0           9.1  8.421453
12481                  The Dark Knight     12269.0           8.3  8.265477
2843                        Fight Club      9678.0           8.3  8.256385
292                       Pulp Fiction      8670.0           8.3  8.251406
522                   Schindler's List      4436.0           8.3  8.206639
23673                         Whiplash      4376.0           8.3  8.205404
5481                     Spirited Away      3968.0           8.3  8.196055
2211                 Life Is Beautiful      3643.0           8.3  8.187171
1178            The Godfather: Part II      3418.0           8.3  8.180076
1152   One Flew Over the Cuckoo's Nest      3001.0           8.3  8.164256
351                       Forrest Gump      8147.0           8.2  8.150272
1154           The Empire Strikes Back      5998.0           8.2  8.132919
1176                            Psycho      2405.0           8.3  8.132715

BREAK THE JSON FILES 

In [41]:
md.head()

belongs_to_collection    budget  \
0  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1                                                NaN  65000000   
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3                                                NaN  16000000   
4  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                                              genres  \
0  [{'id': 16, 'name': 'Animation'}, {'id': 35, '...   
1  [{'id': 12, 'name': 'Adventure'}, {'id': 14, '...   
2  [{'id': 10749, 'name': 'Romance'}, {'id': 35, ...   
3  [{'id': 35, 'name': 'Comedy'}, {'id': 18, 'nam...   
4                     [{'id': 35, 'name': 'Comedy'}]   

                               homepage     id    imdb_id original_language  \
0  http://toystory.disney.com/toy-story    862  tt0114709                en   
1                                   NaN   8844  tt0113497                en   
2                                   NaN  15602  tt0113228                en   
3                                   NaN  31357  tt0114885                en   
4                                   NaN  11862  tt0113041                en   

                original_title  \
0                    Toy Story   
1                      Jumanji   
2             Grumpier Old Men   
3            Waiting to Exhale   
4  Father of the Bride Part II   

                                            overview popularity  ...  \
0  Led by Woody, Andy's toys live happily in his ...  21.946943  ...   
1  When siblings Judy and Peter discover an encha...  17.015539  ...   
2  A family wedding reignites the ancient feud be...    11.7129  ...   
3  Cheated on, mistreated and stepped on, the wom...   3.859495  ...   
4  Just when George Banks has recovered from his ...   8.387519  ...   

                                production_countries release_date  \
0  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-10-30   
1  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-15   
2  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
3  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-12-22   
4  [{'iso_3166_1': 'US', 'name': 'United States o...   1995-02-10   

       revenue runtime                                   spoken_languages  \
0  373554033.0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]   
1  262797249.0   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
2          0.0   101.0           [{'iso_639_1': 'en', 'name': 'English'}]   
3   81452156.0   127.0           [{'iso_639_1': 'en', 'name': 'English'}]   
4   76578911.0   106.0           [{'iso_639_1': 'en', 'name': 'English'}]   

     status                        title  video vote_average vote_count  
0  Released                    Toy Story  False          7.7     5415.0  
1  Released                      Jumanji  False          6.9     2413.0  
2  Released             Grumpier Old Men  False          6.5       92.0  
3  Released            Waiting to Exhale  False          6.1       34.0  
4  Released  Father of the Bride Part II  False          5.7      173.0  

[5 rows x 22 columns]

In [199]:
md.sort_values(by=['id'], ascending=True)

adult belongs_to_collection   budget  \
2429   False                   NaN  1350000   
13609  False                   NaN        0   
4435   False                   NaN        0   
17451  False                   NaN        0   
36946  False                   NaN        0   
...      ...                   ...      ...   
25652  False                   NaN        0   
3767   False                   NaN        0   
12549  False                   NaN        0   
25079  False                   NaN        0   
13106  False                   NaN        0   

                                           genres  \
2429                              [Comedy, Crime]   
13609                             [Comedy, Drama]   
4435                    [Comedy, Science Fiction]   
17451                                [Drama, War]   
36946                                     [Drama]   
...                                           ...   
25652                                    [Comedy]   
3767                       [Action, Crime, Drama]   
12549  [Fantasy, Horror, Action, Science Fiction]   
25079                                    [Comedy]   
13106                              [Crime, Drama]   

                                                homepage      id    imdb_id  \
2429   http://www.universalstudiosentertainment.com/l...     100  tt0120735   
13609                                                NaN   10000  tt0109747   
4435                                                 NaN   10001  tt0096486   
17451                                                NaN  100010  tt0032477   
36946                                                NaN  100017  tt0488903   
...                                                  ...     ...        ...   
25652                                                NaN   99946  tt0016832   
3767                                                 NaN    9995  tt0216772   
12549                                                NaN    9997  tt0857376   
25079                                                NaN   99977  tt0079308   
13106                      http://www.derfreiewille.com/    9999  tt0499101   

      original_language                       original_title  \
2429                 en  Lock, Stock and Two Smoking Barrels   
13609                es            La estrategia del caracol   
4435                 en                       Young Einstein   
17451                en                       Flight Command   
36946                de                             Verfolgt   
...                 ...                                  ...   
25652                en                         Exit Smiling   
3767                 en                           Turn It Up   
12549                ab                              Gabriel   
25079                en                            Hot Stuff   
13106                de                      Der freie Wille   

                                                overview  ...    revenue  \
2429   A card sharp and his unwillingly-enlisted frie...  ...  3897569.0   
13609  A group of tenants living in an old house are ...  ...        0.0   
4435   Albert Einstein is the son of a Tasmanian appl...  ...        0.0   
17451  A rookie flyer, Ens. Alan Drake, joins the fam...  ...        0.0   
36946  Deals with the obsessive relationship between ...  ...        0.0   
...                                                  ...  ...        ...   
25652  Beatrice Lillie and Jack Pickford star in this...  ...        0.0   
3767   Trying to bootstrap his way out of Brooklyn's ...  ...        0.0   
12549  Gabriel tells the story of an archangel who fi...  ...        0.0   
25079  When a police department's burglary task force...  ...        0.0   
13106  After nine years in psychiatric detention Theo...  ...        0.0   

      runtime                                   spoken_languages    status  \
2429    105.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
13609   116.0     

In [56]:
from ast import literal_eval

In [57]:
md['genres'] = md['genres'].fillna('[]').apply(literal_eval).apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])


In [60]:
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   
3      [Comedy, Drama, Romance]                                   NaN  31357   
4                      [Comedy]                                   NaN  11862   

     imdb_id original_language               original_title  \
0  tt0114709                en                    Toy Story   
1  tt0113497                en                      Jumanji   
2  tt0113228                en             Grumpier Old Men   
3  tt0114885                en            Waiting to Exhale   
4  tt0113041                en  Father of the Bride Part II   

                                            overview  ...      revenue  \
0  Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1  When siblings Judy and Peter discover an encha...  ...  262797249.0   
2  A family wedding reignites the ancient feud be...  ...          0.0   
3  Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4  Just when George Banks has recovered from his ...  ...   76578911.0   

  runtime                                   spoken_languages    status  \
0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2   101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3   127.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
4   106.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  year  
0                    Toy Story  False          7.7     5415.0  1995  
1                      Jumanji  False          6.9     2413.0  1995  
2             Grumpier Old Men  False          6.5       92.0  1995  
3            Waiting to Exhale  False          6.1       34.0  1995  
4  Father of the Bride Part II  False          5.7      173.0  1995  

[5 rows x 25 columns]

In [59]:
md['year'] = pd.to_datetime(md['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [61]:
md.head()

adult                              belongs_to_collection    budget  \
0  False  {'id': 10194, 'name': 'Toy Story Collection', ...  30000000   
1  False                                                NaN  65000000   
2  False  {'id': 119050, 'name': 'Grumpy Old Men Collect...         0   
3  False                                                NaN  16000000   
4  False  {'id': 96871, 'name': 'Father of the Bride Col...         0   

                         genres                              homepage     id  \
0   [Animation, Comedy, Family]  http://toystory.disney.com/toy-story    862   
1  [Adventure, Fantasy, Family]                                   NaN   8844   
2             [Romance, Comedy]                                   NaN  15602   
3      [Comedy, Drama, Romance]                                   NaN  31357   
4                      [Comedy]                                   NaN  11862   

     imdb_id original_language               original_title  \
0  tt0114709                en                    Toy Story   
1  tt0113497                en                      Jumanji   
2  tt0113228                en             Grumpier Old Men   
3  tt0114885                en            Waiting to Exhale   
4  tt0113041                en  Father of the Bride Part II   

                                            overview  ...      revenue  \
0  Led by Woody, Andy's toys live happily in his ...  ...  373554033.0   
1  When siblings Judy and Peter discover an encha...  ...  262797249.0   
2  A family wedding reignites the ancient feud be...  ...          0.0   
3  Cheated on, mistreated and stepped on, the wom...  ...   81452156.0   
4  Just when George Banks has recovered from his ...  ...   76578911.0   

  runtime                                   spoken_languages    status  \
0    81.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
1   104.0  [{'iso_639_1': 'en', 'name': 'English'}, {'iso...  Released   
2   101.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
3   127.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   
4   106.0           [{'iso_639_1': 'en', 'name': 'English'}]  Released   

                                             tagline  \
0                                                NaN   
1          Roll the dice and unleash the excitement!   
2  Still Yelling. Still Fighting. Still Ready for...   
3  Friends are the people who let you be yourself...   
4  Just When His World Is Back To Normal... He's ...   

                         title  video vote_average vote_count  year  
0                    Toy Story  False          7.7     5415.0  1995  
1                      Jumanji  False          6.9     2413.0  1995  
2             Grumpier Old Men  False          6.5       92.0  1995  
3            Waiting to Exhale  False          6.1       34.0  1995  
4  Father of the Bride Part II  False          5.7      173.0  1995  

[5 rows x 25 columns]

In [164]:
### Qulified - when vote count is greater than min no of votes and not null AND Vote Average is not null 
qualified = md[(md['vote_count'] >= m) & (md['vote_count'].notnull()) & (md['vote_average'].notnull())][['id','title', 'year', 'vote_count', 'vote_average', 'popularity', 'genres', 'belongs_to_collection','homepage','imdb_id','original_title', 'overview', 'poster_path', 'status','video','original_language']]
qualified['vote_count'] = qualified['vote_count'].astype('int')
qualified['vote_average'] = qualified['vote_average'].astype('int')
qualified.shape

(4555, 16)

In [165]:
qualified

id                        title  year  vote_count  vote_average  \
0         862                    Toy Story  1995        5415             7   
1        8844                      Jumanji  1995        2413             6   
4       11862  Father of the Bride Part II  1995         173             5   
5         949                         Heat  1995        1886             7   
8        9091                 Sudden Death  1995         174             5   
...       ...                          ...   ...         ...           ...   
45177  375798                      Brice 3  2016         160             4   
45204  417870                   Girls Trip  2017         393             7   
45258  417320                Descendants 2  2017         171             7   
45265  265189                Force Majeure  2014         255             6   
45343  248705   The Visitors: Bastille Day  2016         167             4   

      popularity                                             genres  \
0      21.946943                        [Animation, Comedy, Family]   
1      17.015539                       [Adventure, Fantasy, Family]   
4       8.387519                                           [Comedy]   
5      17.924927                   [Action, Crime, Drama, Thriller]   
8        5.23158                      [Action, Adventure, Thriller]   
...          ...                                                ...   
45177   7.276407                                           [Comedy]   
45204  37.964872                                           [Comedy]   
45258  15.842073  [TV Movie, Family, Action, Comedy, Music, Adve...   
45265  12.165685                                    [Comedy, Drama]   
45343    7.29492                                           [Comedy]   

                                   belongs_to_collection  \
0      {'id': 10194, 'name': 'Toy Story Collection', ...   
1                                                    NaN   
4      {'id': 96871, 'name': 'Father of the Bride Col...   
5                                                    NaN   
8                                                    NaN   
...                                                  ...   
45177  {'id': 442352, 'name': 'Brice Collection', 'po...   
45204                                                NaN   
45258  {'id': 466463, 'name': 'Descendants Collection...   
45265                                                NaN   
45343  {'id': 52888, 'name': 'The Visitors Collection...   

                                   homepage    imdb_id  \
0      http://toystory.disney.com/toy-story  tt0114709   
1                                       NaN  tt0113497   
4                                       NaN  tt0113041   
5                                       NaN  tt0113277   
8                                       NaN  tt0114576   
...                                     ...        ...   
45177                                   NaN  tt5029602   
45204                                   NaN  tt3564472   
45258                                   NaN  tt5117876   
45265                                   NaN  tt2121382   
45343                                   NaN  tt2441982   

                     original_title  \
0                         Toy Story   
1                           Jumanji   
4       Father of the Bride Part II   
5                              Heat   
8                      Sudden Death   
...                             ...   
45177                       Brice 3   
45204                    Girls Trip   
45258                 Descendants 2   
45265                        Turist   
45343  Les Visiteurs: La Révolution   

                                                overview  \
0      Led by Woody, Andy's toys live happily in his ...   
1      When siblings Judy and Peter discover an encha...   
4      Just when George Banks has recovered from his ...   
5      Obsessive master thief, Neil McCauley leads a ...   
8      International action supers

Therefore, to qualify to be considered for the chart, a movie has to have at least 434 votes on TMDB. We also see that the average rating for a movie on TMDB is 5.6 on a scale of 10. 4555 Movies qualify to be on our chart.

In [166]:
qualified['Weighted Rating'] = qualified.apply(weighted_rating, axis=1)

In [196]:
qualified = qualified.sort_values('id', ascending=True)

In [197]:
qualified.head()

id            title  year  vote_count  vote_average popularity  \
17     5       Four Rooms  1995         539             6   9.026586   
256   11        Star Wars  1977        6778             8  42.149697   
6232  12     Finding Nemo  2003        6292             7  25.497794   
351   13     Forrest Gump  1994        8147             8  48.307194   
2742  14  American Beauty  1999        3438             7  20.726578   

                                    genres  \
17                         [Crime, Comedy]   
256   [Adventure, Action, Science Fiction]   
6232                   [Animation, Family]   
351               [Comedy, Drama, Romance]   
2742                               [Drama]   

                                  belongs_to_collection  \
17                                                  NaN   
256   {'id': 10, 'name': 'Star Wars Collection', 'po...   
6232  {'id': 137697, 'name': 'Finding Nemo Collectio...   
351                                                 NaN   
2742                                                NaN   

                                               homepage    imdb_id  \
17                                                  NaN  tt0113101   
256   http://www.starwars.com/films/star-wars-episod...  tt0076759   
6232              http://movies.disney.com/finding-nemo  tt0266543   
351                                                 NaN  tt0109830   
2742                      http://www.dreamworks.com/ab/  tt0169547   

       original_title                                           overview  \
17         Four Rooms  It's Ted the Bellhop's first night on the job....   
256         Star Wars  Princess Leia is captured and held hostage by ...   
6232     Finding Nemo  Nemo, an adventurous young clownfish, is unexp...   
351      Forrest Gump  A man with a low IQ has accomplished great thi...   
2742  American Beauty  Lester Burnham, a depressed suburban father in...   

                           poster_path    status  video original_language  \
17    /eQs5hh9rxrk1m4xHsIz1w11Ngqb.jpg  Released  False                en   
256   /btTdmkgIvOi0FFip1sPuZI2oQG6.jpg  Released  False                en   
6232  /syPWyeeqzTQIxjIUaIFI7d0TyEY.jpg  Released  False                en   
351   /yE5d3BUhE8hCnkMUJOo1QDoOGNz.jpg  Released  False                en   
2742  /or1MP8BZIAjqWYxPdPX724ydKar.jpg  Released  False                en   

      Weighted Rating  
17           5.912608  
256          7.945073  
6232         6.965734  
351          7.954125  
2742         6.938553

The top movie according to our metrics is Bollywood's Dilwale Dulhania Le Jayenge

In [169]:
# calculate the average rating by movieID

rating_average = rating.groupby(['movieId']).mean()[['rating']]

In [170]:
rating_average

rating
movieId          
1        3.888157
2        3.236953
3        3.175550
4        2.875713
5        3.079565
...           ...
176267   4.000000
176269   3.500000
176271   5.000000
176273   1.000000
176275   3.000000

[45115 rows x 1 columns]

MERGE DATASET INTO ONE 

In [171]:
## Merging  Credits with Movies Qualified Data (Taking credits with id and merge with Movie id ) = MERGE1
qualified['id'] = qualified.id.astype(int)


In [172]:
merge1 = qualified.merge(credits,on='id')

In [173]:
merge1.head()

id                        title  year  vote_count  vote_average  \
0   19404  Dilwale Dulhania Le Jayenge  1995         661             9   
1   27205                    Inception  2010       14075             8   
2     155              The Dark Knight  2008       12269             8   
3  157336                 Interstellar  2014       11187             8   
4     550                   Fight Club  1999        9678             8   

   popularity                                             genres  \
0   34.457024                           [Comedy, Drama, Romance]   
1   29.108149  [Action, Thriller, Science Fiction, Mystery, A...   
2  123.167259                   [Drama, Action, Crime, Thriller]   
3   32.213481                [Adventure, Drama, Science Fiction]   
4   63.869599                                            [Drama]   

                               belongs_to_collection  \
0                                                NaN   
1                                                NaN   
2  {'id': 263, 'name': 'The Dark Knight Collectio...   
3                                                NaN   
4                                                NaN   

                                       homepage    imdb_id  \
0                                           NaN  tt0112870   
1         http://inceptionmovie.warnerbros.com/  tt1375666   
2  http://thedarkknight.warnerbros.com/dvdsite/  tt0468569   
3             http://www.interstellarmovie.net/  tt0816692   
4    http://www.foxmovies.com/movies/fight-club  tt0137523   

                original_title  \
0  Dilwale Dulhania Le Jayenge   
1                    Inception   
2              The Dark Knight   
3                 Interstellar   
4                   Fight Club   

                                            overview  \
0  Raj is a rich, carefree, happy-go-lucky second...   
1  Cobb, a skilled thief who commits corporate es...   
2  Batman raises the stakes in his war on crime. ...   
3  Interstellar chronicles the adventures of a gr...   
4  A ticking-time-bomb insomniac and a slippery s...   

                        poster_path    status  video original_language  \
0  /2gvbZMtV1Zsl7FedJa5ysbpBx2G.jpg  Released  False                hi   
1  /qmDpIHrmpJINaRKAfWQfftjCdyi.jpg  Released  False                en   
2  /1hRoyzDtpgMU7Dz4JF22RANzQO7.jpg  Released  False                en   
3  /nBNZadXqJSdt05SHLqgT0HuC5Gm.jpg  Released  False                en   
4  /adw6Lq9FiC9zjYEpOqfq03ituwp.jpg  Released  False                en   

   Weighted Rating                                               cast  \
0         8.340942  [{'cast_id': 1, 'character': 'Raj Malhotra', '...   
1         7.973229  [{'cast_id': 1, 'character': 'Dom Cobb', 'cred...   
2         7.969339  [{'cast_id': 35, 'character': 'Bruce Wayne / B...   
3         7.966415  [{'cast_id': 9, 'character': 'Joseph Cooper', ...   
4         7.961264  [{'cast_id': 4, 'character': 'The Narrator', '...   

                                                crew  
0  [{'credit_id': '57a3054a9251417c57000d7a', 'de...  
1  [{'credit_id': '56e8462cc3a368408400354c', 'de...  
2  [{'credit_id': '55a0eb4a925141296b0010f8', 'de...  
3  [{'credit_id': '54cba75b925141678e014d1a', 'de...  
4  [{'credit_id': '55731b8192514111610027d7', 'de...

In [174]:
## Merging Merge1(i.e. Qualified Data and Credits) with Keywords = MERGE 2

merge2 = merge1.merge(keywords,on='id')

In [175]:
merge2.head()

id                        title  year  vote_count  vote_average  \
0   19404  Dilwale Dulhania Le Jayenge  1995         661             9   
1   27205                    Inception  2010       14075             8   
2     155              The Dark Knight  2008       12269             8   
3  157336                 Interstellar  2014       11187             8   
4     550                   Fight Club  1999        9678             8   

   popularity                                             genres  \
0   34.457024                           [Comedy, Drama, Romance]   
1   29.108149  [Action, Thriller, Science Fiction, Mystery, A...   
2  123.167259                   [Drama, Action, Crime, Thriller]   
3   32.213481                [Adventure, Drama, Science Fiction]   
4   63.869599                                            [Drama]   

                               belongs_to_collection  \
0                                                NaN   
1                                                NaN   
2  {'id': 263, 'name': 'The Dark Knight Collectio...   
3                                                NaN   
4                                                NaN   

                                       homepage    imdb_id  \
0                                           NaN  tt0112870   
1         http://inceptionmovie.warnerbros.com/  tt1375666   
2  http://thedarkknight.warnerbros.com/dvdsite/  tt0468569   
3             http://www.interstellarmovie.net/  tt0816692   
4    http://www.foxmovies.com/movies/fight-club  tt0137523   

                original_title  \
0  Dilwale Dulhania Le Jayenge   
1                    Inception   
2              The Dark Knight   
3                 Interstellar   
4                   Fight Club   

                                            overview  \
0  Raj is a rich, carefree, happy-go-lucky second...   
1  Cobb, a skilled thief who commits corporate es...   
2  Batman raises the stakes in his war on crime. ...   
3  Interstellar chronicles the adventures of a gr...   
4  A ticking-time-bomb insomniac and a slippery s...   

                        poster_path    status  video original_language  \
0  /2gvbZMtV1Zsl7FedJa5ysbpBx2G.jpg  Released  False                hi   
1  /qmDpIHrmpJINaRKAfWQfftjCdyi.jpg  Released  False                en   
2  /1hRoyzDtpgMU7Dz4JF22RANzQO7.jpg  Released  False                en   
3  /nBNZadXqJSdt05SHLqgT0HuC5Gm.jpg  Released  False                en   
4  /adw6Lq9FiC9zjYEpOqfq03ituwp.jpg  Released  False                en   

   Weighted Rating                                               cast  \
0         8.340942  [{'cast_id': 1, 'character': 'Raj Malhotra', '...   
1         7.973229  [{'cast_id': 1, 'character': 'Dom Cobb', 'cred...   
2         7.969339  [{'cast_id': 35, 'character': 'Bruce Wayne / B...   
3         7.966415  [{'cast_id': 9, 'character': 'Joseph Cooper', ...   
4         7.961264  [{'cast_id': 4, 'character': 'The Narrator', '...   

                                                crew  \
0  [{'credit_id': '57a3054a9251417c57000d7a', 'de...   
1  [{'credit_id': '56e8462cc3a368408400354c', 'de...   
2  [{'credit_id': '55a0eb4a925141296b0010f8', 'de...   
3  [{'credit_id': '54cba75b925141678e014d1a', 'de...   
4  [{'credit_id': '55731b8192514111610027d7', 'de...   

                                            keywords  
0                  [{'id': 4344, 'name': 'musical'}]  
1  [{'id': 1014, 'name': 'loss of lover'}, {'id':...  
2  [{'id': 849, 'name': 'dc comics'}, {'id': 853,...  
3  [{'id': 83, 'name': 'saving the world'}, {'id'...  
4  [{'id': 825, 'name': 'support group'}, {'id': ...

In [235]:
### Merging Merge2 (i.e Qualified Data , Credits, Keywords) with Links = MERGE 3 

merge3 = pd.merge(merge2, links.set_index('tmdbId'), left_on='id', right_index=True)

In [236]:
merge3.head()

id                        title  year  vote_count  vote_average  \
0   19404  Dilwale Dulhania Le Jayenge  1995         661             9   
1   27205                    Inception  2010       14075             8   
2     155              The Dark Knight  2008       12269             8   
3  157336                 Interstellar  2014       11187             8   
4     550                   Fight Club  1999        9678             8   

   popularity                                             genres  \
0   34.457024                           [Comedy, Drama, Romance]   
1   29.108149  [Action, Thriller, Science Fiction, Mystery, A...   
2  123.167259                   [Drama, Action, Crime, Thriller]   
3   32.213481                [Adventure, Drama, Science Fiction]   
4   63.869599                                            [Drama]   

                               belongs_to_collection  \
0                                                NaN   
1                                                NaN   
2  {'id': 263, 'name': 'The Dark Knight Collectio...   
3                                                NaN   
4                                                NaN   

                                       homepage    imdb_id  ...  \
0                                           NaN  tt0112870  ...   
1         http://inceptionmovie.warnerbros.com/  tt1375666  ...   
2  http://thedarkknight.warnerbros.com/dvdsite/  tt0468569  ...   
3             http://www.interstellarmovie.net/  tt0816692  ...   
4    http://www.foxmovies.com/movies/fight-club  tt0137523  ...   

                        poster_path    status  video original_language  \
0  /2gvbZMtV1Zsl7FedJa5ysbpBx2G.jpg  Released  False                hi   
1  /qmDpIHrmpJINaRKAfWQfftjCdyi.jpg  Released  False                en   
2  /1hRoyzDtpgMU7Dz4JF22RANzQO7.jpg  Released  False                en   
3  /nBNZadXqJSdt05SHLqgT0HuC5Gm.jpg  Released  False                en   
4  /adw6Lq9FiC9zjYEpOqfq03ituwp.jpg  Released  False                en   

  Weighted Rating                                               cast  \
0        8.340942  [{'cast_id': 1, 'character': 'Raj Malhotra', '...   
1        7.973229  [{'cast_id': 1, 'character': 'Dom Cobb', 'cred...   
2        7.969339  [{'cast_id': 35, 'character': 'Bruce Wayne / B...   
3        7.966415  [{'cast_id': 9, 'character': 'Joseph Cooper', ...   
4        7.961264  [{'cast_id': 4, 'character': 'The Narrator', '...   

                                                crew  \
0  [{'credit_id': '57a3054a9251417c57000d7a', 'de...   
1  [{'credit_id': '56e8462cc3a368408400354c', 'de...   
2  [{'credit_id': '55a0eb4a925141296b0010f8', 'de...   
3  [{'credit_id': '54cba75b925141678e014d1a', 'de...   
4  [{'credit_id': '55731b8192514111610027d7', 'de...   

                                            keywords movieId   imdbId  
0                  [{'id': 4344, 'name': 'musical'}]   36086   112870  
1  [{'id': 1014, 'name': 'loss of lover'}, {'id':...   79132  1375666  
2  [{'id': 849, 'name': 'dc comics'}, {'id': 853,...   58559   468569  
3  [{'id': 83, 'name': 'saving the world'}, {'id'...  109487   816692  
4  [{'id': 825, 'name': 'support group'}, {'id': ...    2959   137523  

[5 rows x 22 columns]

In [ ]:
### Merging Merge3 i.e (Qualified Data , Credits, Keywords, links) with ratings = MERGE4


merge4 = merge3.merge(rating,on='movieId')

In [ ]:
merge4.head()

In [233]:
merge4.sort_values(by=['movieId'], ascending=True)

ValueError: 'movieId' is both an index level and a column label, which is ambiguous.